allow slurm to be used as scheduler

adamcallison opened the pull request:
the other commit just moves the part that checks if the scheduler is valid to the top of gen_qsub_script

adamcallison commented:
One issue with this is that it could be confusing that a `slurm` submission is invoked by `grow_qsub` (and similar for `gen_qsub_script`) as `slurm` uses `sbatch` instead of qsub, but I thought an entirely new function was overkill as most of the logic is the same. Considering renaming these as `grow_cluster` and `gen_cluster_script` or similar, and just providing `grow_qsub` and `gen_qsub_script` as thin wrappers so nothing breaks. Thoughts?

adamcallison commented:
Also, this has been (as usual with this `qsub` stuff) difficulto test. I have used this to run stuff on a `slurm` cluster but I've not tried all the options.

jcmgray commented:
Great! Yes they have slurm here and I was planning to add this soon but glad to be beaten to it ;)

I think probably whats makes sense is yes: ``grow_cluster`` and ``gen_cluster_script``, with the ``scheduler`` argument no longer optional. Then
```python
crop.grow_sge
crop.grow_pbs
crop.glow_slurm
```
for convenience & ideally a deprecation warning for using the now slightly ambiguous ``qsub_grow``. What do you think?

I can add these things if you don't fancy doing them in this PR!

pep8speaks commented:
Hello @adamcallison! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`xyzpy/gen/batch.py`](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py):

> [Line 1421:1](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1421): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 1424:5](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1424): [E125](https://duckduckgo.com/?q=pep8%20E125) continuation line with same indent as next logical line
> [Line 1440:5](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1440): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1441:5](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1441): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1444:1](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1444): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 1447:5](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1447): [E125](https://duckduckgo.com/?q=pep8%20E125) continuation line with same indent as next logical line
> [Line 1463:5](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1463): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1464:5](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1464): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1486:1](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1486): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 1489:1](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1489): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1
> [Line 1490:64](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1490): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 1490:66](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1490): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 1492:1](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1492): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 1495:1](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1495): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1
> [Line 1496:64](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1496): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 1496:66](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1496): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 1498:1](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1498): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 1501:1](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1501): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1
> [Line 1502:64](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1502): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 1502:66](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1502): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 1522:1](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1522): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 1525:1](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1525): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1
> [Line 1526:58](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1526): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 1526:60](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1526): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 1528:1](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1528): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 1531:1](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1531): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1
> [Line 1532:58](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1532): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 1532:60](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1532): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 1534:1](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1534): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 1537:1](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1537): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1
> [Line 1538:58](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1538): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 1538:60](https://github.com/jcmgray/xyzpy/blob/60bd561e29fa4bac086e36144a84eb458fbabf6e/xyzpy/gen/batch.py#L1538): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals

adamcallison commented:
Forgot to do the pep8 check. Will do that now.

I implemented what we discussed. I added `DeprecationWarning`s where appropriate but it seems that they are ignored by default in python. Do you know what the correct approach is here?

adamcallison commented:
Also, I think it's worth you taking a look at how I've done the `extra_resources` arg for `slurm`. (Either before accepting the PR or once you start using it). It works fine but could be handled differently - it's not clear to me what the right way is.

adamcallison commented on xyzpy/gen/batch.py:
```diff
@@ -1373,18 +1404,154 @@ def qsub_grow(
+
+def grow_sge(crop, batch_ids=None, **kwargs):  # pragma: no cover
+    grow_cluster(crop, 'sge', batch_ids=batch_ids, **kwargs)
+
```
I thought about doing this, but didn't for 2 reaons:

1) I think giving `scheduler` to `partial` as a keyword cause an error if `batch_ids` is given as a positional arg?

2) The docstring will be wrong, as it will include the `scheduler` arg

jcmgray replied:
1. I don't think so (and have just tried a toy example), think its just like calling ``fn(*arg, **kwargs, scheduler='sge')``
2. The docs function signature should show the arg has already been set, I think better to have one extra arg like this in the docstring than having to look up most args in a different function.

```python
In [1]: import functools

In [2]: def foo(a, *, b):
    ...:     """Foo docs.
    ...:     """
    ...:     return a + b
    ...:

In [3]: foo2 = functools.partial(foo, b=2)

In [4]: functools.update_wrapper(foo2, foo)
Out[4]: functools.partial(<function foo at 0x7fda74eab950>, b=2)

In [5]: foo2(1)
Out[5]: 3

In [6]: foo2?
Signature:       foo2(a, *, b)
Call signature:  foo2(*args, **kwargs)
Type:            partial
String form:     functools.partial(<function foo at 0x7fda74eab950>, b=2)
File:            ~/<ipython-input-30-8b71f7e4af39>
Docstring:
Foo docs.

Class docstring:
partial(func, *args, **keywords) - new function with partial application
of the given arguments and keywords.
```

adamcallison replied:
I was convinced - I tried
```python
grow_slurm = functools.partial(grow_cluster, scheduler='slurm')
functools.update_wrapper(grow_slurm, grow_cluster)
Crop.grow_slurm = grow_slurm
```
but when I tried to call it with 
```python
c = xyz.Crop(name=f'mb_32', parent_dir='/ddn/data/fscs35/xyzpy_m2sbnb2')
c.grow_slurm(batch_ids=(2,), hours=24, gigabytes=4, temp_gigabytes=0) 
```
I get
```python
TypeError: grow_cluster() missing 1 required positional argument: 'crop'
```
It seems that assigning a `partial` object as an attribute to a class doesn't make it get treated like an instance method in the way that it does with a function.

jcmgray replied:
mm this is unknown territory for me. Maybe it needs to be a separate ``functools.partialmethod`` wrap? I've always wondered why this was needed. E.g.
```python
Crop.grow_slurm = functools.partialmethod(Crop.grow_cluster, scheduler='slurm')
```

adamcallison replied:
Cool, I've not seen that before, but from reading the docs it sounds like it should work. I'll try tomorrow.

jcmgray replied:
It seems to on a simple ex:
```python
import functools

class Foo:
     def __init__(self, x):
         self.x = x

def add(f, a, b):
    return f.x + a + b


Foo.add = add
Foo.add3 = functools.partialmethod(Foo.add, b=3)

f = Foo(1)
f.add(2, 3) == f.add3(2) == 6
# True
```
thanks, sorry this has ended up being more complex than I thought!

adamcallison replied:
After some digging, I've found out that the reason it doesn't work is because, unlike `function`, `partial` doesn't adhere to the 'descriptor protocol' which allows a `function` attribute of instances to turn into an `instancemethod` object when its name is resolved. The `partialmethod` object does implement the protocol, which means that `partialmethod` objects get turned into `partial` objects when the name gets resolved.

Unfortunately, `partialmethod` objects _must_ be used as methods: they can't be directly called which means we'd have to make a `partial` version for the module-level functions and `partialmethod` versions for the `Crop` methods. Also, `update_wrapper` seems to do nothing on `partialmethod` objects - at least, I've tried all variations I can think of and it doesn't seem to work.

A much less complicated thing would be to do this:
```python
def grow_slurm(*args, **kwargs):
    return grow_cluster(*args, **{**{'scheduler':'slurm'}, **kwargs})
functools.update_wrapper(grow_slurm, grow_cluster)
Crop.grow_slurm = grow_slurm
```
which works fine for the module-level version and the `Crop` method, including the updated wrapper, but the docs won't give any hint that one argument has been fixed.

What do you think?

jcmgray replied:
The plot thickens! I do think it should be pretty clear from the context that scheduler has been set... but I also think maybe the answer is just to remove the module level functions, and only have methods / partialmethods.

AdrianSosic commented:
Hi @adamcallison and @jcmgray, I am currently also considering to use xyzpy in combination with slurm and therefore just wanted to ask if you are still actively working on this feature (since there haven't been any further comments in the last few months). Kind regards!

adamcallison commented:
@AdrianSosic and @jcmgray  I had some difficulties getting the `partialmethod` objects to document themselves, since `functools.update_wrapper` doesn't seem to work on them, so I never finished this properly. I've just added a commit that removes the module level `grow_sge` etc functions and just adds them to `Crop` as `partialmethod` objects and doesn't attempt to add a documentation wrapper because I can't figure out how to do it. The commit should work in terms of functionality (although would benefit from more 'in the wild' testing. @jcmgray what do you think?

jcmgray commented:
Yeah all looks good to me! Sorry its taken me a while to get back to this, but I am looking forward to trying it. Thanks for the contribution @adamcallison. @AdrianSosic do let us know if you try it and have any problems - this kind of functionality is hard to test.

AdrianSosic commented:
Great, thanks for adding the commit! Appreciate! I will try it out and definitely give you feedback if I encounter any problems.
